rfcs/2023-01-06-layout.md

```diff
@@ -0,0 +1,460 @@
+
+Here's a proposed timeline:
+
+1. `Box` component + responsive styles
+1. Layout styling props on components
+
+## Drawbacks/Constraints
+
+- `className` and `style` props aren't supported. A drawback of this is people relying on maintaining their styles separately (`.css` files) wouldn't be able to set custom CSS properties on blade components. They'll have to rely on the supported primitives and components. As a constraint this helps us to keep the components less prone to getting styled unexpectedly _(a `padding` property on CSS can mess up the built-in spacing on a component)_. For example - exposing `className` enables users to [style the component](https://styled-components.com/docs/basics#styling-any-component) to anything.
+- If someone is customizing the spacing theme tokens then default spacing typings might be off. This might need a utility such as [this](https://chakra-ui.com/docs/styled-system/advanced-theming#theme-typings) to correctly generate typings if we see usecases in the future.
+- People can pass values that aren't part of spacing tokens eg. `7px`. It was considered to keep these restricted only to spacing tokens but that would have heavily restricted genuine usecases where some off scale values might be needed eg. in landing pages.
+- We're not accepting `number` values for styling props and properties in order to avoid confusions. However, this behavior is different from vanilla React where you could pass unit-less values and they're internally transformed to corresponding pixel counterparts.
+- We've constrained the supported properties both on layout components plus styling props in order to avoid unexpected usage, look and feel of components. This might feel too restrictive to some users compared to other general purpose design systems. The idea is to iterate and evolve what all properties should be supported based on feedback and usecases we see as users start using these primitives.
+- Minor caveat of object based responsive properties, users will pass object literals so it would re-render all the time, however not a concern for now. This is applicable in general to React.
+- A drawback of not exposing the helper components such as `Flex`, `Grid`, `Spacer` from our side is it can lead to deviating implementations. Some users may directly use `Box` component everywhere which will make the markup slightly harder to read (compared to having more readable `Flex`, `Grid`, etc.) in the JSX. However, as a constraint this ensures we don't have multiple components to do the same thing, affecting learning curve and usage.
+
+As a contrasting philosophy, Primer has an interesting take on supporting all properties _which means you can actually style a button to look something very different from a button_:
+
+> Assume that people will break the rules, provide safe ways for them to do so
```

Comment:
Reminded me of this 😄 

![css](https://user-images.githubusercontent.com/2913308/215753048-9281ef32-dfe8-42dd-9c1c-46384b6652ab.gif)